### DISCUSSION TITLE
Can I use this in my ionic 3.x configuration?

### DISCUSSION THREAD

**Discussion by AlGantori** [kolbasa/cordova-plugin-apkupdater#43] [Q&A] [answered]
Will this plugin work with my 3.x configuration below
I noticed my "Cordova Platforms" says "not available" I am able to browser and APKs builds but cannot move to Ionic 4,5,or 6 :) too much change I can handle.

![image](https://user-images.githubusercontent.com/11699920/166843351-d88dc1e9-3414-40cf-8fa1-767b2dda95dd.png)

I cannot see from your sample code where it checks if the update is needed or not? How can I do that?
![image](https://user-images.githubusercontent.com/11699920/166843503-58a953d8-8e95-4381-98c9-66b03024936f.png)

Which of these installation steps should I attempt?
![image](https://user-images.githubusercontent.com/11699920/166844732-fb1ad341-ba91-4510-b466-81d6eefd3621.png)

Thank you.

**Comment by kolbasa** [comment, marked as answer]
> Will this plugin work with my 3.x configuration below

The plugin should work with all Ionic versions. I myself still work with Ionic v1.

In your case, you probably need both plugins:
```
ionic cordova plugin add cordova-androidx-build
ionic cordova plugin add cordova-plugin-apkupdater
```

If you update cordova from 9 to 11, then you don't need `cordova-androidx-build` anymore. I'll have to update my readme on this.
```
npm i -g cordova
```

---

> I noticed my "Cordova Platforms" says "not available"

This is normal. This has no effect on the plugin. You can easily add a new platform with:
```
ionic cordova platform add android
```

---

> I cannot see from your sample code where it checks if the update is needed or not? How can I do that?

This is done on purpose. The plugin itself does not provide a version comparison option.
[In the readme](https://github.com/kolbasa/cordova-plugin-apkupdater#update-versioning) I have an extra section with a practical [example](https://github.com/kolbasa/cordova-plugin-apkupdater/blob/master/doc/sample/Versioning.md).

Feel free to write if there are more questions. 
The plugin is actively kept alive by me.

**Comment by AlGantori** [comment]
Thank you for the step by step help.
I have went ahead and issued the following CLI commands...
![image](https://user-images.githubusercontent.com/11699920/167235963-8c9f6524-e39e-4ba7-ad65-c2ce28c7f69c.png)

Skipped (ionic cordova plugin add cordova-androidx-build) per your instruction (I like less dependencies too)

While the Cordova CLI is now at 11 the Cordova Platforms: Android is only at 7 it looks like your plugin requires 9.

![image](https://user-images.githubusercontent.com/11699920/167235919-383cc90b-1b4d-4735-876a-062b38acb74e.png)

I went ahead and factored out your provided code from app.components.ts into an APKUpdater service


I call  this.APKUpdater.Check() from app.component.ts constructor.

![image](https://user-images.githubusercontent.com/11699920/167237249-bdcc9b5e-d87f-4307-a53e-5afdcca7e5b2.png)

The project builds OK, I am able to make an APK as before.

I copied your apk-update.js script into my project but I had to install some missing packages...

```
npm install archiver
npm install node-apk
```
which revealed a bunch (160+) of vulnerabilities I guess in my pakage.json configuration (nothing new, I made several attempts to fix in the past however my skills with package.json are poor).

I think "JS folks" would install your script as a peer project with its own node_module?  correct?

Calling the script as below finally worked and generates 2 files:

```
./src/apk-update.js  S:/_W/ARP/UPS.ion/SRC/UPS.ion.3/platforms/android/app/build/outputs/apk/release/app-release.apk  S:/_W/ARP/UPS.ion/DEPLOY/UPDATES
```

![image](https://user-images.githubusercontent.com/11699920/167236600-37e3ccdf-4838-49d7-b32b-725a9f7c5949.png)

I am not sure if your plugin derives "app-release.zip" from the json "name": "app-release.apk"  in order to download it as it's not specified anywhere?

Can the plugin/script simply use app-release.apk from the build without wrapping it into a .zip? 

I uploaded the 2 produced files as such to my server

![image](https://user-images.githubusercontent.com/11699920/167236528-0ffaa536-507e-44c1-8a73-2aa4489b5873.png)

I incremented the APK version and repeated the build process to ensure I am running an APK with the apkupdater plugin integrated however it does not look like the .Update() code is doing anything.  I was hoping to get a "visual debug feedback (fail/success)" from one of the popups I used in APKUpdater.Update() method but I get neither.

I reviewed some of my steps and I noticed after my your plugin install step

```
ionic cordova plugin add cordova-plugin-apkupdater
```
it said...

```
Installing "cordova-plugin-apkupdater" for android
Plugin doesn't support this project's cordova-android version. cordova-android: 7.0.0, failed version requirement: >=9.0.0
Skipping 'cordova-plugin-apkupdater' for android
Installing "cordova-plugin-apkupdater" for browser
Adding cordova-plugin-apkupdater to package.json
Android Studio project detected  (I ONLY INSTALLED THIS FOR THE SDK ADB TOOLS).

```
I don't know how to if I can bump up cordova-android to 9 or higher?

![image](https://user-images.githubusercontent.com/11699920/167237517-00327cbb-120f-4454-83e3-2b3d497509a2.png)

tried   ionic cordova platform add android@9 --verbose

![image](https://user-images.githubusercontent.com/11699920/167239712-6816c479-98ac-4c32-be4a-559c4313f75b.png)


Sorry for the length but I thought it's better to give you more info about my "guess" work.
Thank you so much!!! 🙏

...
![image](https://user-images.githubusercontent.com/11699920/167238601-0e509971-6fc2-4f89-ba1c-c88fe6ba7745.png)

**Comment by AlGantori** [comment]
Update...
decided to try remove/add

ionic cordova platform remove android
ionic cordova platform add android

![image](https://user-images.githubusercontent.com/11699920/167243019-40f11ecf-91c4-439a-8e2f-181efe93ec18.png)

Cordova Platforms: android is at 10.1.2

![image](https://user-images.githubusercontent.com/11699920/167243043-0e273b6b-ee40-484a-a02d-6ef54927332c.png)

ionic cordova plugin add cordova-plugin-apkupdater
Success !?
![image](https://user-images.githubusercontent.com/11699920/167243053-090cbc72-7d14-44f6-8294-9d9959cec0dd.png)

**Build APK is now broken**

cordova build android --release -- --keystore="..." --storePassword= ....

![image](https://user-images.githubusercontent.com/11699920/167243240-4822baef-6651-4d90-b7ed-c26c36e02fff.png)

It looks like it has switched to using gradle tool?  I know only the name "gradle"...

I fixed my SDK environment variable in .bash_profile added new ANDROID_SDK_ROOT=

![image](https://user-images.githubusercontent.com/11699920/167243168-a3d7f589-c49a-400c-9f8c-5e93238b8701.png)

but it's still failing any ideas?

![image](https://user-images.githubusercontent.com/11699920/167243293-163dd7ea-975e-490e-8d2b-fb4d6fdb9e77.png)

I run SDK manager via Android Studio, however, I don't see a 30.0.3 version of the tools only 30.0.1 which is already installed.

![image](https://user-images.githubusercontent.com/11699920/167243574-892b2985-f943-437d-97dd-1925db08a03a.png)

![image](https://user-images.githubusercontent.com/11699920/167243695-b052724d-416d-461a-b7f9-1133f6019f45.png)

**Comment by AlGantori** [comment]
More Updates...
I managed to find the "SDK build tools" under show "Package Details"

![image](https://user-images.githubusercontent.com/11699920/167243989-185c14c5-5d25-4a18-9282-50f1809fc239.png)

Build APK succeeds but produces a app-release.aab ??
aab = Android App Bundle

![image](https://user-images.githubusercontent.com/11699920/167244279-df3ad105-e90f-411e-9eea-30699b2642aa.png)

I tried installing this .aab on this Android 11 box but it does not like it

![image](https://user-images.githubusercontent.com/11699920/167244425-f0aa2606-896b-4365-9f5f-3c8cdfb616ce.png)

**Comment by kolbasa** [comment]
You are on the right track. It's a pain, but it's worth keeping `cordova` and `cordova-android` up to date.

To generate an apk, you need to adjust your command a bit.
```
--packageType=apk
```

Complete example:
```
ionic cordova build android --release -- -- --packageType=apk --verbose
```

---

> I copied your apk-update.js script into my project which revealed a bunch (160+) of vulnerabilities I guess in my pakage.json configuration (nothing new, I made several attempts to fix in the past however my skills with package.json are poor)

You can safely ignore all these warnings. These are local dev libraries on your computer. They don't lead to security problems in your app itself.
What can lead to problems is if you don't keep`cordova` and `cordova-android` up to date.

---

> I think "JS folks" would install your script as a peer project with its own node_module? correct?

`apk-update` is meant to be a global node script, so it can be used via the terminal.

That's why it says `-g` in the instructions.
```
npm install -g apk-update
```

---

> Can the plugin/script simply use app-release.apk from the build without wrapping it into a .zip?

Yes, the plugin can also download an .apk file directly. 
The `apk-update` script is just an example of how I do it. I like it compressed. You can ignore the script completely.

You can write your own solution as it suits you best.
The plugin gives you the possibility to read which app version is currently installed. 
Now you just need to somehow query your server if there is a new version. This can be a simple text file with the current version in it, that's basically the way I go.

Maybe you have a remote api that can tell if there is a new version.

**Comment by AlGantori** [comment]
Thank you so much! for the support.  I am now able to produce APK as usual.  Down down GooglePlay 
Earlier inorder to get a visual feedback which APK version I was running I decided to display the APK version in my AppLogoComp so I installed the AppVersion plugin only to discover your plugin also provides this information.
I have now commented that code out and probably should remove the AppVersion plugin altogether.

While the APK is building I am getting white screen or incomplete network access.  What do you use to debug an APK and be able to see console log output?  I don't remember anymore if Genymotion several years ago allowed me to do that?  It certainly gave me a headache this year with all its new install requirements, I failed.  For now I am trying to see if running in browser gives me any clues...

Do you have any idea why await ApkUpdater.getInstalledVersion() is failling saying ApkUpdater is "undefined"

I am not familiar with the import notation (ApkUpdater is outside the {} braces )
![image](https://user-images.githubusercontent.com/11699920/167250921-40d49092-d841-4ed4-9f09-a1058c335349.png)
does that mean ApkUpdater it is a class with static members that does not need to be injected as a service?

![image](https://user-images.githubusercontent.com/11699920/167250953-3bbcda48-61ef-499b-b790-6d8c35db7580.png)

![image](https://user-images.githubusercontent.com/11699920/167250957-e67a5c74-1287-4c04-8b2f-f8a259eee66f.png)

I get zero errors in the console the moment I comment out the call to my ApkUpdaterService.Update()

The remote revision is getting retrieved fine from the remote json 0.0.18 mapped into 18 

![image](https://user-images.githubusercontent.com/11699920/167262902-6180ba58-051c-49de-9e72-d48aa43f9297.png)

polyfills.js:3 TypeError: Cannot read properties of undefined (reading 'getInstalledVersion')

Stuck 🙃

**Comment by kolbasa** [comment]
Sorry, could not reply for a while.

When installing a plugin, the best way is actually to test the code first via Chrome DevTools. 
1. Install the app on an emulator/device.
2. Open `chrome://inspect/#devices` with Google Chrome.
3. Inspect your App
4. Type `ApkUpdater` and press Enter.

All functions of the plugin should be displayed now. If they are not shown, then the plugin was not installed correctly.

![DevTools](https://user-images.githubusercontent.com/3901582/167294030-c8b3274d-7fd4-4c73-a93c-930f01161ad8.gif)


Commands from the video above.
```js
await ApkUpdater.download('https://raw.githubusercontent.com/kolbasa/cordova-plugin-apkupdater-demo/master/Demo.apk', {onDownloadProgress: console.log});
await ApkUpdater.install();
```

Alternatively, as a classic Promise notation:
```js
ApkUpdater.download('https://raw.githubusercontent.com/kolbasa/cordova-plugin-apkupdater-demo/master/Demo.apk', {onDownloadProgress: console.log}).then(ApkUpdater.install)
```

Alternatively with callbacks.
```js
ApkUpdater.download('https://raw.githubusercontent.com/kolbasa/cordova-plugin-apkupdater-demo/master/Demo.apk', {onDownloadProgress: console.log}, ApkUpdater.install)
```

---

> I am not familiar with the import notation (ApkUpdater is outside the {} braces )

The ApkUpdater-module is simply the default export  and therefore can be imported without {} braces..

The plugin can be accessed in the global window object. You can ignore my Typescript api and go directly via `window.ApkUpdater` if that is easier.

---

**Reply by AlGantori** [reply]
With the new configuration needed for ApkUpdater  some of my interface (images, etc) where not loading however calls to an Image API over https are working.  So it was until very late last night that I dug up some of my old HTML notes on "Remote Debugging". I was amazed I could see the console from the APK running on the Android 11 device (a TV Box).

So it looks like with the new APK builds, they now load the application page index.html via https: causing a "Mixed Content" error that was not the case before.  However, I don't see the "Mixed Content" errors now when running in the browser, puzzling 🙃.


![image](https://user-images.githubusercontent.com/11699920/167308818-a279e858-4b3b-411a-8e5f-ae33059ea9ea.png)


Thank you for your directions calling ApkUpdater from the console it does **Fail** in the browser perhaps that's expected? 

 but it **Succeeds** when running in the device as an APK.

From app running in Browser...

![image](https://user-images.githubusercontent.com/11699920/167308851-c19956ea-d5fe-471e-945a-c862a5380371.png)

From APK on device...

![image](https://user-images.githubusercontent.com/11699920/167308870-a7deb452-f666-4b1c-a3ee-95286080a98c.png)

When installing the plugin it said it installed for both Android and Browser? so I assumed it will handle both platforms? 
In other words when running APKUpdater in Browser it would stay silent (provides no functionality).

Please confirm if I should be calling ApkUpdater with conditional code like

```
 if (this.Platform.platforms().toString().Contains("android")) {  
      ApkUpdater... 
 } ?
```

Because of the "Mixed Content" errors, I guess I cannot continue anymore without an SSL, so I am going to have to get an SSL certificate first to regain normal loading of my assets in the APK mode, then I will re-enable (uncomment) the code calling ApkUpdater.

🙏

![image](https://user-images.githubusercontent.com/11699920/167309297-357ff7eb-791b-47aa-8ec9-3119b5dcd680.png)

**Reply by kolbasa** [reply]
> it does Fail in the browser

Yes, the plugin is not available in the browser. I'll see if I can include it in the next release.
Here I would just check if it runs on the device or not.
```js
if (!this.platform.is('android')) {
    return;
}
```

---

> Because of the "Mixed Content" errors, I guess I cannot continue anymore without an SSL

I can't help you there. This is probably an after-effect of the Cordova upgrade. You should have a look into the Cordova documentation.  Maybe there is a setting that can be adjusted. 

Here is a discussion about it:
https://forum.ionicframework.com/t/allowing-mixed-content-on-android-cordova-app/208943/10
There are quite a few proposed solutions.

**Reply by AlGantori** [reply]
Hello, so I have managed to get my site secure via cloudflare SSL certificate and after 3 days of fights and waiting, I am now back at integrating and testing ApkUpdater.  

Here is what I am observing.,..

(1) The app-release.json will get cached I assume by the browser remote on my pc or the APK's webview, I have no idea.
Is there a way to force download without requiring a Ctrl-F5 on the remote debug browser?
I say this because for several runs my debug lines would say remoteVersion 18 and installedVersion is 18 as well.

So I finally got 19
![image](https://user-images.githubusercontent.com/11699920/168409382-211e79c8-6115-4570-ad1c-494654669575.png)

(2) but no it does not look like it updated the app

![image](https://user-images.githubusercontent.com/11699920/168409400-152f0706-560d-4d60-8822-1efa805f0c81.png)

I tried ApkUpdater.download("https://mumti.org/UPDATES/app-release.apk") on the console

![image](https://user-images.githubusercontent.com/11699920/168409409-cb5a69da-852f-44f4-b29c-740223f66665.png)

(3) now I see the "manually" downloaded .apk file in the storage 

![image](https://user-images.githubusercontent.com/11699920/168409419-4b98d31a-8c5b-4dfc-8157-2795593d7bb8.png)

(4) To debug further how should I call apk-update.js to skip generating the .zip file and simply I guess tell it copy over the app-release.apk I point to from the lastest build. (which I upload with a free-filesync script)

(5) I tried top "manually" download  the .zip file but I does NOT seem to make it into the device storage

![image](https://user-images.githubusercontent.com/11699920/168409635-33e2bd1d-8da0-4949-bbc2-8867b93f1185.png)

(6) my current ionic/cordova config versions (ratatouille)

```
Ionic:

Ionic CLI : 6.13.1 (C:\Users\AXM\AppData\Roaming\npm\node_modules\@ionic\cli)
Ionic Framework : ionic-angular 3.9.2
@ionic/app-scripts : 3.2.4

Cordova:

Cordova CLI : 11.0.0
Cordova Platforms : android 10.1.2, browser 5.0.3
Cordova Plugins : cordova-plugin-ionic-webview 4.1.3, (and 15 other plugins)

Utility:

cordova-res : 0.15.3
native-run : not installed

System:

Android SDK Tools : 26.1.1 (C:/Users/AXM/AppData/Local/Android/Sdk)
NodeJS : v10.16.0 (C:\Program Files\nodejs\node.exe)
npm : 6.9.0
OS : Windows 10
```

(6)
Earlier when it was stuck on the cache I was getting the Popup() from the else now the code is entering the if-then but somehow failing in the download or install ???  So I have added my Log.Write() to see how far it's getting in the console.

![image](https://user-images.githubusercontent.com/11699920/168409762-80f80dae-f871-4bb9-82fc-a47182d3f7a1.png)

Your insights much much appreciated.  Thank you.

**Reply by kolbasa** [reply]
> Is there a way to force download without requiring a Ctrl-F5 on the remote debug browser?

This can be configured on your server, depending on what you use there. 
A popular trick is to add the timestamp to the URL. This is often sufficient.

```js
let UpdateInfoURL = Globals.AppSiteRoot + '/UPDATES/app-release.ison?' + Date.now();
```

---

You are passing the version file to the download method here and not the zip or apk file:

![image](https://user-images.githubusercontent.com/3901582/168443651-32f9fa05-4d13-497b-8453-87bbc1c939f6.png)

That is better:

![image](https://user-images.githubusercontent.com/3901582/168443852-a18f575e-7a77-47f2-b197-19ddf78e23b6.png)

Zip-File download:
```js
await ApkUpdater.download(
    'https://mumti.org/UPDATES/app-release.zip',
    {
        onDownloadProgress: console.log,
        onUnzipProgress: console.log
    }
);
await ApkUpdater.install();
```

Apk-file download:
```js
await ApkUpdater.download(
    'https://mumti.org/UPDATES/app-release.apk',
    {
        onDownloadProgress: console.log
    }
);
await ApkUpdater.install();
```

![Mumti](https://user-images.githubusercontent.com/3901582/168444633-144abb7c-15db-49f3-bce3-e6c6a8ba8eac.gif)

`ApkUpdater.download()` by the way also provides detailed infos of the downloaded file.

```js
const update = await ApkUpdater.download(
    'https://mumti.org/UPDATES/app-release.apk',
    {
        onDownloadProgress: console.log
    }
);
console.log(update);
```
-->
```json
{
  "name": "app-release.apk",
  "path": "/data/user/0/io.ionic.starter/files/update",
  "size": 9378331,
  "checksum": "f98e0a3196c568ed675836bab7509518",
  "app": {
    "name": "MUMTI G3",
    "package": "org.mumti.MUMTI.G3",
    "version": {
      "code": 19,
      "name": "0.0.19"
    }
  }
}
```

**Comment by AlGantori** [comment]
Finally Success!!!

![image](https://user-images.githubusercontent.com/11699920/168451903-26ed0485-9ecb-4499-b909-8d594bc04efb.png)

![image](https://user-images.githubusercontent.com/11699920/168451670-53083e0a-46ac-4753-9c62-7b76e2bf3460.png)

Thank you for pointing out all my mistakes, I was not paying attention and made the wrong conclusion that .download() was taken the json info as input, that's the reason I had asked earlier,
       >>> I am not sure if your plugin derives "app-release.zip" from the json "name": "app-release.apk" in order 
       >>> to download it as it's not specified anywhere?

But it's totally my mistake I should have read and followed more carefully your provided example.

I get the following popup on the device and I think that outside of ApkUpdater control, but something I would to dig around the device if it is possible to make the update totally silent (without user interventions).

![image](https://user-images.githubusercontent.com/11699920/168451032-eed2f4af-0720-40cd-8d03-8c450020a032.png)

Thanks for showing me I needed an await at the console prompt I am certainly poor at what can be done at the dev console (I am from the era of TurboPascal, then brainwashed with all the Microsoft garbage of C++, C# and a thousand never-ending frameworks).

I had to drop almost all of that in order to get on the web/android with Node.js / Ionic / Angular / Cordova

Perhaps, you may be German?, about 50 years ago I visited Würzburg, I was 10 then, my brother was working there for Hoechst, Beautiful country. May I ask what tools you use for capturing screen recording and turning them into animated .gifs?

Thanks for all the assistance and guidance. 
Now that I have the **"BLD Automatic APK Updates"** task/log completed.  I still have few things to work on like:

1) Revisiting why my service-worker.js + increment_version_service-worker.js, is not  causing my web builds (PWA) to auto download the updates to (build/assets/...) like it used to, couple months ago.

2) I am just starting to play with some REST API against MongoDB in order to migrate my > 5,300 files of XML metadata.

**Reply by kolbasa** [reply]
> I get the following popup on the device and I think that outside of ApkUpdater control, but something I would to dig around the device if it is possible to make the update totally silent (without user interventions).

It is possible to install this automatically without the dialog. 

You have to root the device or register the app as the device owner.
I have an extra section on this in my readme. 

https://github.com/kolbasa/cordova-plugin-apkupdater#rootinstall
and
https://github.com/kolbasa/cordova-plugin-apkupdater#ownerinstall

However, this requires some work on the device itself. 

It would then look like this. Fully automated:

![OwnerInstall](https://user-images.githubusercontent.com/3901582/168465050-cedbf600-61bb-4122-beba-a5c60d9c82b2.gif)

---

> Perhaps, you may be German?, about 50 years ago I visited Würzburg

Yes, I live in Germany and even in said Würzburg :smile:
Where are you from?

---

>  then brainwashed with all the Microsoft garbage of C++, C# and a thousand never-ending frameworks

Welcome to JavaScript in the new world of never-ending frameworks :smile:

---

> May I ask what tools you use for capturing screen recording and turning them into animated .gifs?

I am on macOS. Unfortunately, the program does not exist under Windows:
https://getkap.co
